### PR TITLE
refactor(themes): update themes build and output to generated

### PR DESCRIPTION
**Pull request by joshblack**
Part 2 of: https://github.com/carbon-design-system/carbon/pull/4061

This PR looks to refactor parts of our themes package to make it a little easier to work with, namely:

- Refactor `tasks/build.js` to be more file oriented. We now build files for tokens, mixins, and themes
- Move to `generated` folder approach, `scss` folder now has us defining the public API and using files from `generated`
- Add support for inlining custom properties with `custom-property` mixin
- Add support for emitting difference with `should-emit` mixin
- Refactor test setup so that tests are split per sass entrypoint file

#### Changelog

**New**

- Add support for `custom-property` and `should-emit` for part 3 of work

**Changed**

- Update public Sass API entrypoints for `themes`
- Refactor test structure for `themes`

**Removed**

### CODE REVIEW

**Comment by joshblack** [comment]
Note to reviewers, spotted a regression for inline theming so will take a look at that 👀

File: packages/components/src/globals/scss/vendor/@carbon/themes/scss/_mixins.scss [hunk 1]
```diff
@@ -1,170 +1,35 @@
...
-    --skeleton-01: #{map-get($theme, 'skeleton-01')};
-    --skeleton-02: #{map-get($theme, 'skeleton-02')};
-    --brand-01: #{map-get($theme, 'brand-01')};
-    --brand-02: #{map-get($theme, 'brand-02')};
-    --brand-03: #{map-get($theme, 'brand-03')};
-    --active-01: #{map-get($theme, 'active-01')};
-    --hover-field: #{map-get($theme, 'hover-field')};
+/// @access private
+/// @group @carbon/themes
+@mixin custom-property($name, $value, $prefix: $custom-property-prefix) {
```
Comment (emyarod) [inline packages/components/src/globals/scss/vendor/@carbon/themes/scss/_mixins.scss:14]: this mixin along with `should-emit()` are also generated from `@carbon/themes` right? so we are separating these two from the `generated` directory?

**Review by emyarod — COMMENTED** [review]
one question about our generated theme files

File: packages/components/src/globals/scss/vendor/@carbon/themes/scss/_mixins.scss (hunk 1 above)
Comment (joshblack) [reply packages/components/src/globals/scss/vendor/@carbon/themes/scss/_mixins.scss:14]: @emyarod these ones are authored in `themes/scss/_mixins.scss` whereas the generated code lives under `themes/scss/generated/*` and is brought in by `themes/scss/_mixins` 👍

**Review by emyarod — APPROVED** [review]
looks good to me

File: packages/themes/src/tokens.js [hunk 2]
```diff
@@ -13,7 +13,6 @@ const colors = [
   'interactive02',
   'interactive03',
   'interactive04',
-  'danger',
```
Comment (mattrosno) [inline packages/themes/src/tokens.js:16]: Was `danger` token recently added upstream? Is this removal intentional?

Comment (mattrosno) [reply packages/themes/src/tokens.js:16]: Oh, maybe add this to the "removed" portion of PR description.

File: packages/themes/scss/_mixins.scss [hunk 3]
```diff
@@ -1,170 +1,35 @@
...
+@function should-emit($theme-a, $theme-b, $token, $emit-difference) {
+  @if $emit-difference == false {
+    @return true;
   }
+  @return map-get($theme-a, $token) != map-get($theme-b, $token);
 }
+
+// We import mixins last so that these methods are made available to the
+// carbon--theme mixin
+@import './generated/mixins';
```
Comment (mattrosno) [inline packages/themes/scss/_mixins.scss:35]: Perfect! Glad you added this so you import mixins from the themes package once and don't have to worry about if you also have to import mixins from a `/generated/mixins` file.

File: packages/themes/scss/_theme-maps.scss [hunk 4]
```diff
@@ -1,358 +1,11 @@
...
-) !default;
-/// Carbon's default theme
-/// @type Map
-/// @access public
-/// @alias carbon--theme--white
-/// @group @carbon/themes
-$carbon--theme: $carbon--theme--white !default;
+
+// Deprecated import path
+// TODO: remove in next major release
```
Comment (mattrosno) [inline packages/themes/scss/_theme-maps.scss]: Maybe go with `// TODO: remove this file in next major release (v11)` or something in case we miss this deprecation and it then hangs around longer than expected?

File: packages/themes/scss/_tokens.scss [hunk 5]
```diff
@@ -1,327 +1,8 @@
-// Code generated by @carbon/themes. DO NOT EDIT.
 //
```
Comment (mattrosno) [inline packages/themes/scss/_tokens.scss:1]: Is it possible to rename the `_tokens.scss` files here to `_variables.scss` without a breaking change? The direction that we want to go: each package has `index.scss` and `package-name.scss` entrypoints that you choose based on your build. When using the package's Sass, imports are deterministic so if you are looking to include a Sass variable without side-effects, you `@import './path/to/variables.scss';` and you don't have to look up if the file name is actually called variables, or tokens, global-tokens, etc.

File: packages/themes/scss/_mixins.scss [hunk 6]
```diff
@@ -1,170 +1,35 @@
...
-  $disabled-03: map-get($theme, 'disabled-03') !global;
-  $highlight: map-get($theme, 'highlight') !global;
-  $skeleton-01: map-get($theme, 'skeleton-01') !global;
-  $skeleton-02: map-get($theme, 'skeleton-02') !global;
-  $brand-01: map-get($theme, 'brand-01') !global;
-  $brand-02: map-get($theme, 'brand-02') !global;
-  $brand-03: map-get($theme, 'brand-03') !global;
-  $active-01: map-get($theme, 'active-01') !global;
-  $hover-field: map-get($theme, 'hover-field') !global;
+$custom-property-prefix: 'cds' !default;
```
Comment (mattrosno) [inline packages/themes/scss/_mixins.scss:10]: Can this go in a `_variables.scss` file?

File: packages/themes/scss/_mixins.scss [hunk 7]
```diff
@@ -1,170 +1,35 @@
...
   }
+}
 
-  @content;
-  // Reset to default theme after apply in content
-  @if $theme != $carbon--theme {
-    @include carbon--theme();
+/// @access private
+/// @group @carbon/themes
+@function should-emit($theme-a, $theme-b, $token, $emit-difference) {
```
Comment (mattrosno) [inline packages/themes/scss/_mixins.scss:26]: Can this go in a `_functions.scss` file?

**Review by mattrosno — COMMENTED** [review]
This looks great!!! Things I didn't review in depth: build, test, bundler, cli changes.

I noticed the theme switcher add-on in this PR isn't working, React storybook. But then, for some reason I don't see it in our production storybook at all so don't know what's up there. Does that storybook add-on need updated with any of these changes?

Most of my feedback is around where the Sass lives, file names. My longest comment below goes into most detail.

File: packages/themes/src/tokens.js (hunk 2 above)
Comment (joshblack) [reply packages/themes/src/tokens.js:16]: It was added, but there was a duplicate. This removes the duplicate (still exists in the list, just further down 👍 )

File: packages/themes/scss/_tokens.scss (hunk 5 above)
Comment (joshblack) [reply packages/themes/scss/_tokens.scss:1]: I'm definitely torn on the file naming scheme by syntax instead of feature, maybe this could be a broaden change across the project?

**Comment by emyarod** [comment]
I'll take a look at the theme switcher following this and #4033, but it should be working on the current `master` branch (storybook appears to be several releases behind?)

File: packages/themes/scss/_mixins.scss (hunk 6 above)
Comment (joshblack) [reply packages/themes/scss/_mixins.scss:10]: I wanted to keep this internal-only for now versus letting folks mess with it, and try and scope it with the mixin.

Separately, this is one reason I'm not a big fan of file naming per syntax instead of feature because functionality wise it'd be great to have these in the same file, otherwise it's hard to know where things are coming from.

File: packages/themes/scss/_mixins.scss (hunk 7 above)
Comment (joshblack) [reply packages/themes/scss/_mixins.scss:26]: Same notes as above, maybe this could be a broader topic change? Thankfully we could swap any of this out in the future so shouldn't have to block anything now.

**Comment by joshblack** [comment]
Let me dig in and see what's going on! 👀

File: packages/themes/scss/_tokens.scss (hunk 5 above)
Comment (mattrosno) [reply packages/themes/scss/_tokens.scss:1]: It needs to be a broad change across all packages (especially components). If you're building one giant Sass file for your app, it doesn't matter because everything has previously been imported and is available. But with that, a poor development experience with slow re-builds of everything Sass.

When styling in a more modular fashion (e.g. CSS modules), each file can be its own Sass build. Great developer experience, but many more imports. "I need to use the breakpoint **mixin** for this, spacing **variables** for that, a theme **variables** here, theme **mixin** there"... deterministic imports really help.

As this PR is only about the themes package, we can defer a file naming scheme until later. My suggestion is based on what I see in the industry and that if we do it now, we don't have to do it later. It doesn't have to happen now if it requires a separate discussion. Your call.

File: packages/themes/scss/_mixins.scss (hunk 6 above)
Comment (mattrosno) [reply packages/themes/scss/_mixins.scss:10]: Ah, it's designated as private. I don't have a preference on where private variables live. Feel free to resolve this convo.

File: packages/themes/scss/_mixins.scss (hunk 7 above)
Comment (mattrosno) [reply packages/themes/scss/_mixins.scss:26]: Ah, it's designated as private. I don't have a preference on where private functions live. Feel free to resolve this convo.

**Comment by joshblack** [comment]
@mattrosno theme switcher should be fixed 👍

File: packages/themes/scss/_theme-maps.scss (hunk 4 above)
Comment (joshblack) [reply packages/themes/scss/_theme-maps.scss]: Down for whatever comment syntax we want to check for! No preference really

**Review by mattrosno — APPROVED** [review]

**Comment by joshblack** [comment]
Thanks @emyarod and @mattrosno! Will wait for netlify to finish just to verify 👀
